src/compiler.jl

```diff
@@ -87,6 +87,343 @@ function null_free_handler(B::LLVM.API.LLVMBuilderRef, ToFree::LLVM.API.LLVMValu
     return C_NULL
 end
 
+dedupargs() = ()
+dedupargs(a, da, args...) = (a, dedupargs(args...)...)
+
+@generated function alloc(tt::Type{T}) where T
+    quote
+        ref::Any = Base.llvmcall(("""
+        declare {}*** @julia.ptls_states()
+        declare noalias nonnull {} addrspace(10)* @julia.gc_alloc_obj(i8*, i64, {} addrspace(10)*)
+        define {} addrspace(10)* @allocate(i64 %size, {} addrspace(10)* %ty) #1 {
+            %1 = call {}*** @julia.ptls_states()
+            %2 = bitcast {}*** %1 to i8*
+            %res = call noalias nonnull {} addrspace(10)* @julia.gc_alloc_obj(i8* %2, i64 %size, {} addrspace(10)* %ty)
```

Comment:
You are not setting the typetag of the allocation. So I would expect `typeof(alloc(Int))` to fail.

Reply:
```
  %9 = call noalias nonnull {}* @jl_gc_pool_alloc(i8* %ptls_i8, i32 1400, i32 16) #1
  %10 = bitcast {}* %9 to i64*
  %11 = getelementptr inbounds i64, i64* %10, i64 -1
  store atomic i64 140278619936176, i64* %11 unordered, align 8

```

where `140278619936176` is `Base.pointer_from_objref(T)`

Reply:
Oh you are calling gc_alloc_obj, right...

Reply:
```
@generated function alloc(tt::Type{T}) where T
           sz = sizeof(T)
           type = reinterpret(Int64, Base.pointer_from_objref(T))
           mod ="""
               declare {}*** @julia.ptls_states()
               declare noalias nonnull {} addrspace(10)* @julia.gc_alloc_obj(i8*, i64, {}*)
               define {} addrspace(10)* @allocate() #1 {
                   %1 = call {}*** @julia.ptls_states()
                   %2 = bitcast {}*** %1 to i8*
                   %res = call noalias nonnull {} addrspace(10)* @julia.gc_alloc_obj(i8* %2, i64 $sz, {}* inttoptr (i64 $type to {}*))
                   ret {} addrspace(10)* %res
               }
               attributes #1 = { alwaysinline }
               """
           quote
               ref = Base.llvmcall(($mod, "allocate"), Any, Tuple{})
               return ref
           end
       end
```